crates/hir/src/source_analyzer.rs

```diff
@@ -600,13 +603,11 @@ fn scope_for_offset(
             .filter(|it| it.value.kind() == SyntaxKind::MACRO_CALL)?;
             Some((source.value.text_range(), scope))
         })
-        // find containing scope
-        .min_by_key(|(expr_range, _scope)| {
-            (
-                !(expr_range.start() <= offset.value && offset.value <= expr_range.end()),
-                expr_range.len(),
-            )
+        .filter(|(expr_range, _scope)| {
+            expr_range.start() <= offset.value && offset.value <= expr_range.end()
```

Comment:
Filter out scopes whose text range does not contain the `offset`, why do we need to take them into account when we are looking for containing scope before?

But after this change, one of the old tests doesn't pass

![image](https://user-images.githubusercontent.com/30254428/174046340-e08f42fd-5387-41df-9449-6cd340be5d72.png)

Then I found that in the above test case, the `offset.file_id` here is a `MarcoFile` , which is not reasonable, the `offset.value` is relative to the original file, the `offset.file_id` should also be the original file. Wrong `offset.file_id` affects the scope of the `offset`, related code: 

https://github.com/rust-lang/rust-analyzer/blob/7ade4d49fc59ec939294897890375a8b86c743e2/crates/hir/src/source_analyzer.rs#L587-L602

That's why I changed lines 91-94. After I changed this, all the tests passed.

Reply:
Sounds reasonable to me, though the same change should be done for `new_for_body` in line 68-71